src/bmson/bms_to_bmson.rs

```diff
@@ -0,0 +1,356 @@
+                })
+            },
+            judge_rank,
+            total: {
+                let total_value = if let Some(total) = self.header.total.as_ref() {
+                    total.to_f64().unwrap_or(100.0)
+                } else {
+                    warnings.push(BmsToBmsonWarning::MissingTotal);
+                    100.0
+                };
+                FinF64::new(total_value).unwrap_or_else(|| {
+                    warnings.push(BmsToBmsonWarning::InvalidTotal);
+                    FinF64::new(100.0).expect("Internal error: 100.0 is not a valid FinF64")
+                })
+            },
+            back_image: self
+                .header
+                .back_bmp
+                .as_ref()
+                .cloned()
+                .map(|path| path.display().to_string()),
+            eyecatch_image: self
+                .header
+                .stage_file
+                .map(|path| path.display().to_string()),
+            title_image: self.header.back_bmp.map(|path| path.display().to_string()),
+            banner_image: self.header.banner.map(|path| path.display().to_string()),
+            preview_music: None,
+            resolution,
+            ln_type: self.header.ln_mode,
+        };
+
+        let (sound_channels, mine_channels, key_channels) = {
+            let path_root = self.notes.wav_path_root.clone().unwrap_or_default();
+            let mut sound_map: HashMap<_, Vec<Note>> = HashMap::new();
+            let mut mine_map: HashMap<_, Vec<MineEvent>> = HashMap::new();
+            let mut key_map: HashMap<_, Vec<KeyEvent>> = HashMap::new();
+            for note in self.notes.all_notes() {
+                let note_lane = note
+                    .kind
+                    .is_playable()
+                    .then_some(
+                        match note.key {
+                            Key::Key1 => 1,
+                            Key::Key2 => 2,
+                            Key::Key3 => 3,
+                            Key::Key4 => 4,
+                            Key::Key5 => 5,
+                            Key::Key6 => 6,
+                            Key::Key7 => 7,
+                            Key::Scratch | Key::FreeZone => 8,
+                            // TODO: Extra key convertion
+                            Key::Key8
+                            | Key::Key9
+                            | Key::Key10
+                            | Key::Key11
+                            | Key::Key12
+                            | Key::Key13
+                            | Key::Key14
+                            | Key::ScratchExtra
+                            | Key::FootPedal => 0,
+                        } + match note.side {
+                            PlayerSide::Player1 => 0,
+                            PlayerSide::Player2 => 8,
+                        },
+                    )
+                    .and_then(|lane| NonZeroU8::new(lane));
```

Comment:
```suggestion
                    .and_then(NonZeroU8::new);
```
This redundant closure can be reduced.